lib/puma/thread_pool.rb

```diff
@@ -298,27 +301,25 @@ def shutdown(timeout=-1)
         # Wait for threads to finish without force shutdown.
         threads.each(&:join)
       else
-        # Wait for threads to finish after n attempts (+timeout+).
-        # If threads are still running, it will forcefully kill them.
-        timeout.times do
-          threads.delete_if do |t|
-            t.join 1
-          end
-
-          if threads.empty?
-            break
-          else
-            sleep 1
+        join = ->(timeout) do
+          start = Time.now.to_f
+          threads.reject! do |t|
+            t.join timeout + start - Time.now.to_f
```

Comment:
I actually had one thought. Should we be using [monotonic time](https://blog.dnsimple.com/2018/03/elapsed-time-with-ruby-the-right-way/)? Or is `#to_f` good enough for us?